cmd/swarm/swarm-smoke-pss/pss.go

```diff
@@ -0,0 +1,364 @@
+
+const (
+	pssModeRaw = iota
+	pssModeAsym
+	pssModeSym
+)
+
+func (m pssMode) String() string {
+	return [...]string{"raw", "asym", "sym"}[m]
+}
+
+type pssJob struct {
+	sender   *pssNode
+	receiver *pssNode
+	mode     pssMode
+	msg      []byte
+}
+
+type pssNode struct {
+	hostIdx        int
+	addr           []byte
+	pubkey         string
+	client         *rpc.Client
+	deregisterFunc func()
+}
+
+type pssSession struct {
+	topic pss.Topic
+	msgC  chan pss.APIMsg
+	nodes []*pssNode
+	jobs  map[string]*pssJob
+}
+
+func pssAsymCheck(ctx *cli.Context) error {
+	return runCheck(pssModeAsym, pssMessageCount, pssMessageSize)
+}
+
+func pssSymCheck(ctx *cli.Context) error {
+	return runCheck(pssModeSym, pssMessageCount, pssMessageSize)
+}
+func pssRawCheck(ctx *cli.Context) error {
+	return runCheck(pssModeRaw, pssMessageCount, pssMessageSize)
+}
+
+func pssAllCheck(ctx *cli.Context) error {
+	gotErr := false
+	if err := pssRawCheck(ctx); err != nil {
+		log.Error("error when running raw tests", "err", err)
+		gotErr = true
+	}
+	if err := pssSymCheck(ctx); err != nil {
+		log.Error("error when running sym tests", "err", err)
+		gotErr = true
+	}
+	if err := pssAsymCheck(ctx); err != nil {
+		log.Error("error when running asym tests", "err", err)
+		gotErr = true
+	}
+	if gotErr {
+		return errors.New("some tests failed")
+	}
+	return nil
+}
+
+func runCheck(mode pssMode, count int, msgSizeBytes int) error {
+	log.Info(fmt.Sprintf("pss.%s test started", mode), "msgCount", count, "msgBytes", msgSizeBytes)
+
+	session := pssSetup()
+
+	defer func() {
+		for _, n := range session.nodes {
+			n.deregisterFunc()
+		}
+	}()
+
+	if len(session.nodes) <= 1 {
+		return errors.New("at least 2 nodes are required to be working")
+	}
+
+	jobs := session.genJobs(count, mode, msgSizeBytes)
+
+	errc := make(chan error)
+	go func() {
+		var failCount, successCount int64
+		t := time.Now()
+		sc, err := session.processJobs(jobs)
+		if err != nil {
+			log.Error("error processing some jobs", "err", err)
+		}
+		successCount = int64(sc)
+		failCount = int64(count - sc)
+
+		metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.msgs.fail", mode), nil).Inc(failCount)
+		metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.msgs.success", mode), nil).Inc(successCount)
+
+		totalTime := time.Since(t)
+
+		metrics.GetOrRegisterResettingTimer(fmt.Sprintf("pss.%s.total-time", mode), nil).Update(totalTime)
+		log.Info(fmt.Sprintf("pss.%s test ended", mode), "time", totalTime, "success", successCount, "failures", failCount)
+
+		if failCount > 0 {
+			errc <- errors.New("some messages were not delivered")
+		} else {
+			errc <- nil
+		}
+	}()
+
+	select {
+	case err := <-errc:
+		if err != nil {
+			metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.fail", mode), nil).Inc(1)
+		}
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.timeout", mode), nil).Inc(1)
+		return fmt.Errorf("timeout after %v sec", timeout)
```

Comment:
Will this simple return free up all resources?

Reply:
I think so. We have the following within the same function: 
```go
	defer func() {
		for _, n := range session.nodes {
			n.deregisterFunc()
		}
	}()
```

`deregisterFunc()` will close all connections/subscriptions.